scipy/sparse/_compressed.py

```diff
@@ -858,31 +855,10 @@ def _setdiag(self, values, k):
         if 0 in self.shape:
             return
 
-        M, N = self.shape
-        broadcast = (values.ndim == 0)
-
-        if k < 0:
-            if broadcast:
-                max_index = min(M + k, N)
-            else:
-                max_index = min(M + k, N, len(values))
-            i = np.arange(max_index, dtype=self.indices.dtype)
-            j = np.arange(max_index, dtype=self.indices.dtype)
-            i -= k
-
-        else:
-            if broadcast:
-                max_index = min(M, N - k)
-            else:
-                max_index = min(M, N - k, len(values))
-            i = np.arange(max_index, dtype=self.indices.dtype)
-            j = np.arange(max_index, dtype=self.indices.dtype)
-            j += k
-
-        if not broadcast:
-            values = values[:len(i)]
-
-        self[i, j] = values
+        coo = self.tocoo()
+        coo._setdiag(values, k)
+        arrays = coo._coo_to_compressed(self._swap)
+        self.indptr, self.indices, self.data, _ = arrays
```

Comment:
This will always make a new copy of the index and data arrays, while (I think) the old version would update those arrays in-place if possible.

This might not be an issue in practice, but it's something to keep in mind.